Divided Screen into Thirds

Migu3lito5 opened the pull request:
Divided the screen into three zones using the tower decider script. This script will take in the mouse click position and decide what tower should shoot depending on the tower's zone it is located in. Therefore if a mouse click happens on the left side of the screen, the left tower should shoot. 

Also updated the tower controller script slightly.

Migu3lito5 committed:
Added A new Script Tower Decider (if you have a better name lmk)

As hopefully the name implies this script, divides the screen into three equal zones. Then if the player clicks on the screen it will make the tower in the respective zone shoot, while the others don't.

Migu3lito5 committed:
Update TowerController.cs script

Instead of doing the shooting in the update function, I decided to make it into another method called Shoot().

This method will be called from the tower decider script once it picks what tower should shoot.

Migu3lito5 committed:
Scene to demonstrate towers scripts

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,32 @@
+
+public class TowerDecider : MonoBehaviour
+{
+    public GameObject leftTurret;
```
```suggestion
    public Turret leftTurret;
```
Can we change these references to be of the Turret components themselves so we don't have to use GetComponent in this script, please?

Also, in light of that, can we change the name of TowerController to either Tower or Turret. If we aren't going to use separate Turret and Tower logic, we can just call it `Tower`

Referring to the tower as turret or vice versa makes the code confusing.

Migu3lito5 replied:
Yup I have made everything a tower. I can see how it can lead to confusion.

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    private void TowerToShoot(GameObject turret)
```
Once the above change has been made, you don't need this function. Just call
```suggestion
    thisTurret.Fire()
```
on your conditional statements in Update()

Migu3lito5 replied:
Done!

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,32 @@
+            //TODO: Add more logic to this algorithm to decide the next best turret to use,
+            //      if the current one has no missiles left
+
+            if (xPosOnClick < THIRD) TowerToShoot(leftTurret);
```
See the below comment for how to change this.

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,32 @@
+    public GameObject middleTurret;
+    public GameObject rightTurret;
+
+    private float THIRD = Screen.width / 3f;
```
My convention on Rider is telling me this should be called:
```suggestion
    private float _third= Screen.width / 3f;
```
If VS is telling you otherwise, then ignore this comment

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerController.cs:
```diff
@@ -4,24 +4,21 @@ public class TowerController : MonoBehaviour
-    {
-        if (Input.GetMouseButtonDown(1))
-        {
-            Vector3 mousePosition = new Vector3(Input.mousePosition.x, Input.mousePosition.y, Camera.main.nearClipPlane);
```
Not sure how nearClipPlane is used here, but I may be unfamiliar with it.

Can we make a private or hidden public reference to the Camera.main at Start() to avoid the expensive calls on this line and below?

Migu3lito5 replied:
Actually we can make the z-axis just a zero. Thus I think that's better for the time being

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerController.cs:
```diff
@@ -4,24 +4,21 @@ public class TowerController : MonoBehaviour
-            Vector3 mousePosition = new Vector3(Input.mousePosition.x, Input.mousePosition.y, Camera.main.nearClipPlane);
-            Vector3 worldPosition = Camera.main.ScreenToWorldPoint(mousePosition);
-            worldPosition.z = transform.position.z;
-            GameObject missileInstance = Instantiate(missilePrefab, new Vector3(transform.position.x, transform.position.y + offset, 0), Quaternion.identity);
```
This Vector3 will be the same each time a missile is fired from this tower. Can we move this Vector3 to a private variable initialized at Start() ?

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerController.cs:
```diff
@@ -4,24 +4,21 @@ public class TowerController : MonoBehaviour
-            }
+        if (missile != null)
+        {
+            missile.speed = 5;
```
This speed is set after instantiation of the object which may call Update() on itself without a set speed value before this conditional statement. Consider initializing this in Missile.cs, but you could keep this set here as well.

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerController.cs:
```diff
@@ -4,24 +4,21 @@ public class TowerController : MonoBehaviour
+        worldPosition.z = transform.position.z;
+        GameObject missileInstance = Instantiate(missilePrefab, new Vector3(transform.position.x, transform.position.y + offset, 0), Quaternion.identity);
+      
+        // Give the missile the attributes it needs to move towarsd its target position
```
```suggestion
        // Give the missile the attributes it needs to move towards its target position
```

ConlynPattison reviewed (changes requested):
Multiple small changes requested in addition to the changing of the TowerController naming and references to improve readability,

Migu3lito5 committed:
Update: Changed File Name

Updated the script file name and class to be just Tower, instead of tower controller. Also made some minor changes to make improvements on the code

Migu3lito5 committed:
Update: Minor changes

Made changes to the tower decider script and a minor one to missile.cs

Migu3lito5 committed:
Update SampleScene.unity

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,31 @@
+            //TODO: Add more logic to this algorithm to decide the next best turret to use,
+            //      if the current one has no missiles left
+
+            if (xPosOnClick < _third) TowerToShoot(leftTower);
+            else if (xPosOnClick > _third && xPosOnClick < (_third * 2f)) TowerToShoot(middleTower);
+            else TowerToShoot(rightTower);
```
This change should be made in addition to removing TowerToShoot()
```suggestion
            if (xPosOnClick < _third) leftTower.Fire();
            else if (xPosOnClick > _third && xPosOnClick < (_third * 2f)) middleTower.Fire();
            else rightTower.Fire();
```

ConlynPattison commented on Seasons Defense/Assets/Scripts/TowerDecider.cs:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    private void TowerToShoot(Tower tower)
+    {
+        tower.Fire();
+    }
```
This will be redundant unless we need to do more than Fire a missile from the TowerDecider (I don't think so, particle effects and such can be done from the Missile script)

ConlynPattison commented on Seasons Defense/Assets/Scripts/Tower.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    public GameObject missilePrefab;
+    public float offset;
+    private Vector3 towerPosition;
```
For the sake of convention and local sanity. Private instance variables should have the following syntax:
```suggestion
    private Vector3 _towerPosition;
```

Migu3lito5 committed:
Changed Scene name

ConlynPattison commented on Seasons Defense/Assets/Scenes/SampleScene.unity:
Unfortunately, this will be a recurring issue with merge conflicts. Can you please change the name of this file to anything that you'd like. Something like MiguelScene or something to that effect. Hopefully that's all that'll be needed.

ConlynPattison replied:
This didn't work. A scene completely new and different to others' would be best for the debugging phase we're in. Will follow up with a better solution.

ConlynPattison commented on Seasons Defense/Assets/Scripts/Missile.cs:
```diff
@@ -10,6 +10,7 @@ public class Missile : MonoBehaviour
     private void Start()
     {
         transform.LookAt(target);
+        speed = 5;
     }
```
I probably said this off earlier, but you can just set the speed outside of Start() and init it in the instance variable declarations.
```suggestion
    public int speed = 5;
    
    private void Start()
    {
        transform.LookAt(target);
    }
```

ConlynPattison commented on Seasons Defense/Assets/Scripts/Tower.cs:
```diff
@@ -0,0 +1,29 @@
+            missile.target = worldPosition;
+        }
+    }
+}
```
The ultimate request. Make sure to add a newline to the end of every script. GitHub and some low level languages will argue with it.
```suggestion
}

```

Migu3lito5 committed:
Changes made
